### PR TITLE
[spirv] Support {Append|Consume}StructuredBuffer

### PR DESCRIPTION
**Pull request by antiagainst**
* Supported the `{Append|Consume}StructureBuffer` type
* Supported `.Append()` and `.Consume()`

`{Append|Consume}StructureBuffer<S>` is implemented as `OpTypeStruct`
with an `OpTypeRuntimeArray` of `S`. And each such array will have
an associated counter, whose type is `OpTypeStruct` of 32-bit integer.
The index for `.Append()`/`.Conume()` is fetched by doing atomic
operations on the associated counter.

### CODE REVIEW

**Comment by antiagainst** [comment]
@ehsannas PTA

**Comment by antiagainst** [comment]
This PR only touches SPIR-V files

File: tools/clang/test/CodeGenSPIRV/vk.binding.register.hlsl [hunk 1]
```diff
@@ -61,6 +61,20 @@ ConstantBuffer<S> myCbuffer3 : register(b2, space3);
 // CHECK-NEXT: OpDecorate %sbuffer2 Binding 6
 RWStructuredBuffer<S> sbuffer2 : register(u6, space1);
 
+    // The counter variable will use the next unassigned number
+// CHECK:      OpDecorate %abuffer DescriptorSet 0
+// CHECK-NEXT: OpDecorate %abuffer Binding 5
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/vk.binding.register.hlsl]: I was under the impression that the (descriptorSet, Binding) pair should uniquely identify the resource we're referring to.
I see (0,5) used for both `abuffer` and `sbuffer1`. Am I missing something?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 2]
```diff
@@ -1508,6 +1508,69 @@ SPIRVEmitter::processStructuredBufferLoad(const CXXMemberCallExpr *expr) {
...
+  if (isAppend) {
+    // For append, we add one to the counter.
+    index = theBuilder.createAtomicIAdd(u32Type, counterPtr, one, zero, one);
+  } else {
+    // For consume, we substract one from the counter. Note that OpAtomicIAdd
+    // returns the value before the addition; so we need to do substraction
+    // again with OpAtomicIAdd's return value.
+    const uint32_t negOne = theBuilder.getConstantInt32(-1);
+    const auto prevIndex =
+        theBuilder.createAtomicIAdd(u32Type, counterPtr, one, zero, negOne);
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: is there a reason you chose not to use `OpAtomicISub` with `one`?

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 3]
```diff
@@ -140,6 +145,18 @@ uint32_t DeclResultIdMapper::createExternVar(const VarDecl *var) {
   resourceVars.emplace_back(id, getResourceBinding(var),
                             var->getAttr<VKBindingAttr>());
 
+  if (isACSBuffer) {
+    // For {Append|Consume}StructuredBuffer, we need to create another variable
+    // for its associated counter.
+    const uint32_t counterType = typeTranslator.getACSBufferCounter();
+    const std::string counterName = "counter.var." + var->getName().str();
+    const uint32_t counterId =
+        theBuilder.addModuleVar(counterType, storageClass, counterName);
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Yes, we definitely need to increase the counter variable. Sadly `Uniform` storage class doesn't mean we cannot write to it. For example, `RWStructuredBuffer` is also in `Uniform` storage class. (Terms in graphics is just misleading for lots of cases.) `UniformConstant` means read-only.

[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: oh! i see... Terms in graphics are heavily overloaded and mean different things in different contexts... Thanks for the clarification

File: tools/clang/test/CodeGenSPIRV/vk.binding.register.hlsl (hunk 1 above)
Comment (antiagainst) [reply tools/clang/test/CodeGenSPIRV/vk.binding.register.hlsl]: Yes, we should have different pairs for different resources. However, the current implementation doesn't differentiate `t`, `s`, `u`, `b` yet. `sbuffer1` is bound to `t5`, while `abuffer` is bound to `u5`.  But we only take the number into consideration so they are all mapped to binding number 5. We need a way to automatically assign non overlapping numbers eventually.

Comment (ehsannas) [reply tools/clang/test/CodeGenSPIRV/vk.binding.register.hlsl]: Ok, makes sense

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 4]
```diff
@@ -144,6 +144,12 @@ class ModuleBuilder {
   uint32_t createBinaryOp(spv::Op op, uint32_t resultType, uint32_t lhs,
                           uint32_t rhs);
 
+  /// \brief Creates an OpAtomicIAdd instruction with the given SPIR-V opcode.
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Good catch. Fixed.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 2 above)
Comment (antiagainst) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Good catch! Somehow it didn't come to me that I can use subtraction. Ugh.

**Pull request merged**